contracts/cw20-atomic-swap/src/contract.rs

```diff
@@ -622,8 +682,121 @@ mod tests {
         );
     }
+
+    #[test]
+    fn test_native_cw20_swap() {
+        let mut deps = mock_dependencies(CANONICAL_LENGTH, &[]);
+
+        // Create the contract
+        let env = mock_env("anyone", &[]);
+        let res = init(&mut deps, env, InitMsg {}).unwrap();
+        assert_eq!(0, res.messages.len());
+
+        // Native side (offer)
+        let native_sender = HumanAddr::from("A_on_X");
+        let native_rcpt = HumanAddr::from("B_on_X");
+        let native_coins = coins(1000, "tokens_native");
+
+        // Create the Native swap offer
+        let native_swap_id = "native_swap".to_string();
+        let create = CreateMsg {
+            id: native_swap_id.clone(),
+            hash: real_hash(),
+            recipient: native_rcpt.clone(),
+            expires: Expiration::AtHeight(123456),
+        };
+        let env = mock_env(&native_sender, &native_coins);
+        let res = handle(&mut deps, env, HandleMsg::Create(create)).unwrap();
+        assert_eq!(0, res.messages.len());
+        assert_eq!(log("action", "create"), res.log[0]);
+
+        // Cw20 side (counter offer (1:1000))
+        let cw20_sender = HumanAddr::from("B_on_Y");
+        let cw20_rcpt = HumanAddr::from("A_on_Y");
+        let cw20_coin = Cw20CoinHuman {
+            address: HumanAddr::from("my_cw20_token"),
+            amount: Uint128(1),
+        };
+
+        // Create the Cw20 side swap counter offer
+        let cw20_swap_id = "cw20_swap".to_string();
+        let create = CreateMsg {
+            id: cw20_swap_id.clone(),
+            hash: real_hash(),
+            recipient: cw20_rcpt.clone(),
+            expires: Expiration::AtHeight(123000),
+        };
+        let receive = Cw20ReceiveMsg {
+            sender: cw20_sender,
+            amount: cw20_coin.amount,
+            msg: Some(to_binary(&HandleMsg::Create(create)).unwrap()),
+        };
+        let token_contract = cw20_coin.address; // TODO: Confirm
+        let env = mock_env(&token_contract, &[]);
+        let res = handle(&mut deps, env, HandleMsg::Receive(receive.clone())).unwrap();
+        assert_eq!(0, res.messages.len());
+        assert_eq!(log("action", "create"), res.log[0]);
+
+        // Somebody (typically, A) releases the swap side on the Cw20 (Y) blockchain,
+        // using her knowledge of the preimage
+        let env = mock_env("somebody", &[]);
+        let res = handle(
+            &mut deps,
+            env,
+            HandleMsg::Release {
+                id: cw20_swap_id.clone(),
+                preimage: preimage(),
+            },
+        )
+        .unwrap();
+        assert_eq!(1, res.messages.len());
+        assert_eq!(log("action", "release"), res.log[0]);
+        assert_eq!(log("id", cw20_swap_id), res.log[1]);
+
+        // Verify the resulting Cw20 transfer message
+        let send_msg = Cw20HandleMsg::Transfer {
+            recipient: cw20_rcpt,
+            amount: cw20_coin.amount,
+        };
+        assert_eq!(
```

Comment:
Yup. I wonder if some of these asserts could be made easier with a few helpers, but not sure